reframing data pipelines (#2)
datajoint/blog: datajoint_workflows.md

> # Workflow Management for DataJoint

Reviewer: What about?

```suggestion
# DataJoint-driven Workflow Management
```

**Duration Target:** 14min read (@160 wpm)

**Release:** Thursday, 2022-03-24

**Title:** Blog Title

## [DRAFT] Intro
- something clever/attention-grabbing

> - soften tone to be open-minded as an opinionated discussion follows

Reviewer: Since we are still working on adding this, I think we should keep these notes until the section is completed.

# DataJoint Roadmap for Workflow Orchestration

**Duration Target:** 14min read (@160 wpm)

**Release:** Thursday, 2022-03-24

**Title:** Blog Title

## [DRAFT] Intro
- something clever/attention-grabbing
- soften tone to be open-minded as an opinionated discussion follows

> 

Reviewer: We still need this section introduce the discussion. These are notes for consideration. I'm currently working on this and will be updated soon.

> A **job** is a fully-specified request to perform a computation or task in a deterministic manner

Reviewer: Prefer if we highlight or quote this line is some way as it is the concise description of what a job is and the lasting impression of what they should take away from this section. I've noticed this pattern quite a bit in blogs I've come across before.

```suggestion
*A **job** is a fully-specified request to perform a computation or task in a deterministic manner.*
```

Jobs should be specified in their entirety so that they can achieve consistent results in a predictable timeframe. Reliability is achieved if results can be reproduced and independently verifed.

> Consider a job that responsible for compiling a `plotly` graph from some input data. For reliable results, we need to answer several questions:

Reviewer: ```suggestion
Consider a job that is responsible for compiling a `plotly` graph from some input data. For reliable results, we need to answer several questions:
```

- What version of `plotly` and all its dependencies will we use?
- What is the input data and the processing parameters?
- What hardware resources should we allocate? CPU? GPU? RAM? What operating system and hardware architecture?
- How can the job be parallelized through multithreading and multiprocessing?

These last questions may not affect the final result but they do affect the run time.  Answering these questions leads to the predictability of computation results and execution times.

> A **workflow** is the formal specification of the dependency relationship between jobs

Reviewer: Curious on why you feel we should remove the `acyclic` description. Think this is an important definition as workflows must terminate. Also, similarly as above.

```suggestion
### Workflows
*A **workflow** is the formal specification of the acyclic relationship between jobs.*
```

Workflows provide instructions for cascading dependent computations or tasks. We can chain jobs together in this way to organize how inputs transform into results through various stages of a workflow. These acyclic relationships are important because they define a specific direction and a finite end to the operation. This is commonly depicted with a directed acyclic graph (DAG).

Continuing with our example, we define two jobs that will consume the generated `plotly` graph: 

- embed the plot within a provided text input (say for a report)
- log the activity to a central service

This workflow has one parent job and two children jobs.

> Then **job orchestration** is the process of delegating jobs to both an available and appropriate computing resource.

Reviewer: Similarly as above.

```suggestion
### Job Orchestration
***Job orchestration** is the process of delegating jobs to both an available and appropriate computing resource.*
```

Think of the job orchestrator layer as a matchmaker of sorts. By listening to a *job queue*, her responsibility is to connect the right job to the right environment; best fit for the job. Notice that the notion of a workflow here really isn't necessary since all we are doing is scheduling and fullfilling computing obligations. This means that as we process the queue, we either need to spawn an environment suitable for the job or wait until there is bandwidth within the landscape (remote or on-premise) to launch it.

Perhaps in our example we'd like for our jobs to be run within a distributed Kubernetes cluster. Jobs could be encapsulated in pods scheduled against nodes that satisfy the hardware requirements.

> **Workflow Management** is an interface (graphical or programatic) that enables workflow specification and job queueing

Reviewer: Similarly as above.

```suggestion
### Workflow Management
***Workflow management** is an interface (graphical or programatic) that enables workflow specification and job queueing.*
```

Returing to our example, using a workflow management tool would be how we might define our workflow's jobs and set the workflow to be triggered once a day. Though largely automated now, we'd be wise if during processing we assign someone to be on-call should intervention be necessary.

A **data pipeline** is somet different conceptually.  
It is the formal specification of the data sources, transformations, and all its intermediate structures.
It also takes the form of a directed acyclic graph of dependencies. 

> In many cases, the data pipeline provides the basis for the *workflows*.

Reviewer: This is too long. We need a short, concise description of a data pipeline that can easily be quoted from this blog. Also, similarly as above. I don't think this captures that data

```suggestion
*A **data pipeline** is the formal specification of the data sources, transformations, and all its intermediate structures that provide the basis for many workflows.*
```

Returing to our example, using a workflow management tool would be how we might define our workflow's jobs and set the workflow to be triggered once a day. Though largely automated now, we'd be wise if during processing we assign someone to be on-call should intervention be necessary.

Now a **data pipeline** is somewhat different conceptually.  
It is the formal specification of the data sources, transformations, and all its intermediate structures.
It also takes the form of a directed acyclic graph of dependencies. 

> In many cases, the data pipeline provides the basis for the *workflows*.

Reviewer: This is too long. We need a short, concise description of a **data pipeline** that can easily be quoted from this blog. Also, similarly as above. I don't think this captures that data

```suggestion
### Data Pipeline
*A **data pipeline** is the formal specification of the data sources, transformations, and all its intermediate structures that provide the basis for many workflows.*
```

- workflow: add pose tracking to videos and update metadata, input: utilize a specific algorithm
- workflow: generate graphs and figures for publication, input: curated list of session data

## DataJoint's Approach

> When we designed the DataJoint framework, there were several key principles we wanted to adhere to that we felt were either missing or reduced in significance in the management of computation at the time; primarily within neuroscience. Our approach is to empower the user to create and manage relational **data pipelines**. If you are not familiar with the DataJoint framework, please see [this](https://datajoint.org) or initial paper.

Reviewer: ```suggestion
When we designed the DataJoint framework, there were several key principles we wanted to adhere to that we felt were either missing or reduced in significance in the management of computation at the time; primarily within neuroscience. Our approach is to empower the user to create and manage relational **data pipelines**. If you are not familiar with the DataJoint framework, please see [https://datajoint.org](https://datajoint.org) and our original [position paper](https://www.biorxiv.org/content/10.1101/031658v1).
```
